x/evm/keeper/state_transition.go

```diff
@@ -0,0 +1,157 @@
+		Debug:     debug,
+	}
+}
+
+// GetHashFn implements vm.GetHashFunc for Ethermint. It handles 3 cases:
+//  1. The requested height matches the current height from context (and thus same epoch number)
+//  2. The requested height is from an previous height from the same chain epoch
+//  3. The requested height is from a height greater than the latest one
+func (k Keeper) GetHashFn() vm.GetHashFunc {
+	return func(height uint64) common.Hash {
+		switch {
+		case k.ctx.BlockHeight() == int64(height):
+			// Case 1: The requested height matches the one from the context so we can retrieve the header
+			// hash directly from the context.
+			return k.cache.blockHash
+
+		case k.ctx.BlockHeight() > int64(height):
+			// Case 2: if the chain is not the current height we need to retrieve the hash from the store for the
+			// current chain epoch. This only applies if the current height is greater than the requested height.
+			return k.GetHeightHash(k.ctx, height)
+
+		default:
+			// Case 3: heights greater than the current one returns an empty hash.
+			return common.Hash{}
+		}
+	}
+}
+
+// TransitionDb runs and attempts to perform a state transition with the given transaction (i.e Message), that will
+// only be persisted to the underlying KVStore if the transaction does not error.
+//
+// Gas tracking
+//
+// Ethereum consumes gas according to the EVM opcodes instead of general reads and writes to store. Because of this, the
+// state transition needs to ignore the SDK gas consumption mechanism defined by the GasKVStore and instead consume the
+// amount of gas used by the VM execution. The amount of gas used is tracked by the EVM and returned in the execution
+// result.
+//
+// Prior to the execution, the starting tx gas meter is saved and replaced with an infinite gas meter in a new context
+// in order to ignore the SDK gas consumption config values (read, write, has, delete).
+// After the execution, the gas used from the message execution will be added to the starting gas consumed, taking into
+// consideration the amount of gas returned. Finally, the context is updated with the EVM gas consumed value prior to
+// returning.
+//
+// For relevant discussion see: https://github.com/cosmos/cosmos-sdk/discussions/9072
+func (k *Keeper) TransitionDb(ctx sdk.Context, msg core.Message) (*types.ExecutionResult, error) {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), types.MetricKeyTransitionDB)
+
+	cfg, found := k.GetChainConfig(ctx)
+	if !found {
+		return nil, types.ErrChainConfigNotFound
+	}
+
+	// transaction gas meter (tracks limit and usage)
+	startingGas := ctx.GasMeter()
+
+	// NOTE: Since CRUD operations on the SDK store consume gasm we need to set up an infinite gas meter so that we only consume
+	// the gas used by the Ethereum message execution.
+	// Not setting the infinite gas meter here would mean that we are incurring in additional gas costs
+	k.ctx = ctx.WithGasMeter(sdk.NewInfiniteGasMeter())
```

Comment:
Since there are two gas meters, one per block and one per transaction, we should make sure it is easy to manage the state of those safely

Reply:
Some random notes about gas:
EVM consumes gas by adding and subtracting balance..
CommitStateDB manages this.
Balance being subtracted is same token denom defined by the EVM denom params.